### PR TITLE
remove non ecs containers and images changes with docker sdk client

### PR DESCRIPTION
**Pull request by adnxn**
### Summary

updating https://github.com/aws/amazon-ecs-agent/pull/1736 with docker sdk client and merging to dev
### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: 

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by adnxn** [comment]
adding @haikuoliu, @yumex93 since reviewed https://github.com/aws/amazon-ecs-agent/pull/1736

File: agent/config/config_test.go [hunk 1]
```diff
@@ -373,8 +374,8 @@ func TestInvalidFormatParseEnvVariableDuration(t *testing.T) {
 
 func TestValidForImagesCleanupExclusion(t *testing.T) {
 	defer setTestRegion()()
-	defer setTestEnv("ECS_IMAGE_CLEANUP_EXCLUDE", "amazonlinux:2,amazonlinux:3")()
-	imagesNotDelete := parseImageCleanupExclusionList("ECS_IMAGE_CLEANUP_EXCLUDE")
+	defer setTestEnv("ECS_NONECS_IMAGE_CLEANUP_EXCLUDE", "amazonlinux:2,amazonlinux:3")()
```
Comment (yunhee-l) [inline agent/config/config_test.go]: ECS_NONECS? I'm assuming first ECS is the prefix for internal/ECS env vars?
Howe about ECS_EXCLUDE_CLEANUP_NONECS_IMAGE

File: agent/engine/docker_image_manager.go [hunk 2]
```diff
@@ -303,14 +309,157 @@ func (imageManager *dockerImageManager) removeUnusedImages(ctx context.Context)
...
+			seelog.Errorf("Error removing Image %s - %v", kv.ImageName, err)
+			continue
+		} else {
+			seelog.Infof("Image removed: %s", kv.ImageName)
+			numImagesAlreadyDelete++
+		}
+	}
+}
+
+func (imageManager *dockerImageManager) nonECSImagesNames(ctx context.Context) []string {
```
Comment (yunhee-l) [inline agent/engine/docker_image_manager.go]: get*?

File: agent/dockerclient/dockerapi/docker_client.go [hunk 3]
```diff
@@ -999,6 +1003,46 @@ func (dg *dockerGoClient) listContainers(ctx context.Context, all bool) ListCont
...
+	}
+}
+
+func (dg *dockerGoClient) listImages(ctx context.Context) ListImagesResponse {
+	client, err := dg.sdkDockerClient()
+	if err != nil {
+		return ListImagesResponse{Error: err}
+	}
+	images, err := client.ImageList(ctx, types.ImageListOptions{
+		All: false,
```
Comment (yunhee-l) [inline agent/dockerclient/dockerapi/docker_client.go]: can you add comment on what All flag signifies? with false, it's not clear what we are ignoring. From docker API:
"All: Show all images (default hides intermediate images)"
So I'm wondering if this is even required.

File: agent/dockerclient/dockerapi/docker_client_test.go [hunk 4]
```diff
@@ -731,7 +736,7 @@ func TestContainerEventsStreamError(t *testing.T) {
 
 	eventsChan := make(chan events.Message, dockerEventBufferSize)
 	errChan := make(chan error)
-	mockDockerSDK.EXPECT().Events(gomock.Any(), gomock.Any()).Return(eventsChan, errChan).Times(2)
+	mockDockerSDK.EXPECT().Events(gomock.Any(), gomock.Any()).Return(eventsChan, errChan).MinTimes(1)
```
Comment (yunhee-l) [inline agent/dockerclient/dockerapi/docker_client_test.go]: for my curiosity, why did you loosen the required call count check?

File: agent/engine/docker_image_manager.go [hunk 5]
```diff
@@ -303,14 +309,157 @@ func (imageManager *dockerImageManager) removeUnusedImages(ctx context.Context)
...
+			seelog.Infof("Image removed: %s", id)
+			numNonECSContainerDeleted++
+		} else {
+			seelog.Errorf("Error removing Image %s - %v", id, err)
+			continue
+		}
+	}
+}
+
+func (imageManager *dockerImageManager) nonECSContainersIDs(ctx context.Context) ([]string, error) {
```
Comment (yunhee-l) [inline agent/engine/docker_image_manager.go]: get*?

File: agent/config/config_test.go (hunk 1 above)
Comment (samuelkarp) [reply agent/config/config_test.go]: How about `ECS_EXCLUDE_UNTRACKED_IMAGE`?  I'd like to get rid of "NONECS".

File: agent/dockerclient/dockerapi/docker_client_test.go (hunk 4 above)
Comment (adnxn) [reply agent/dockerclient/dockerapi/docker_client_test.go]: so looking at this test `TestContainerEventsStreamError` we only pass up a single event to test this codepath. so we should we expecting at least one call to `Events(...)`.

where as with the `Times(2)`, we would sometimes hit the `Events(...)` call on the first iteration through this [loop](https://github.com/aws/amazon-ecs-agent/blob/dev/agent/dockerclient/dockerapi/docker_client.go#L864) and would get test failures for missing calls to the mock from time to time. for example running with a `-count=100` you'll likely see test failure every time. 

let me know if you think i've missed something -  i think we added this test as part of the moby migration yea?

File: agent/config/config_test.go (hunk 1 above)
Comment (adnxn) [reply agent/config/config_test.go]: yikes, good points. 

> ECS_EXCLUDE_UNTRACKED_IMAGE

i like this, thanks @samuelkarp

File: agent/engine/docker_image_manager.go [hunk 6]
```diff
@@ -303,14 +309,157 @@ func (imageManager *dockerImageManager) removeUnusedImages(ctx context.Context)
...
+}
+
+func (imageManager *dockerImageManager) removeNonECSContainers(ctx context.Context) {
+	nonECSContainersIDs, err := imageManager.nonECSContainersIDs(ctx)
+	if err != nil {
+		seelog.Errorf("Error getting non-ECS container IDs: %v", err)
+	}
+	var nonECSContainerRemoveAvailableIDs []string
+	for _, id := range nonECSContainersIDs {
+		response, _ := imageManager.client.InspectContainer(ctx, id, dockerclient.InspectContainerTimeout)
```
[the remark this answers is not in the document]
Comment (adnxn) [reply agent/engine/docker_image_manager.go]: + checking and logging

File: agent/engine/docker_image_manager.go [hunk 7]
```diff
@@ -303,14 +309,157 @@ func (imageManager *dockerImageManager) removeUnusedImages(ctx context.Context)
...
+	var nonECSImageNamesRemoveEligible []string
+	for _, nonECSImage := range nonECSImageNames {
+		if !isInExclusionList(nonECSImage, imageManager.imageCleanupExclusionList) {
+			nonECSImageNamesRemoveEligible = append(nonECSImageNamesRemoveEligible, nonECSImage)
+		}
+	}
+
+	var imageWithSizeList []ImageWithSize
+	for _, imageName := range nonECSImageNamesRemoveEligible {
+		resp, _ := imageManager.client.InspectImage(imageName)
```
[the remark this answers is not in the document]
Comment (adnxn) [reply agent/engine/docker_image_manager.go]: added checking and logging

**Comment by adnxn** [comment]
```
--- FAIL: TestOOMContainer (10.50s)
	utils_windows.go:87: Registry location test_path_1545881813823392500
	utils_windows.go:97: datadir C:\Users\ADMINI~1\AppData\Local\Temp\amazon-ecs-agent\ftest_temp\ecs_integ_testdata336277954\data
	utils_windows.go:105: Created directory C:\Users\ADMINI~1\AppData\Local\Temp\amazon-ecs-agent\ftest_temp\ecs_integ_testdata336277954 to store test data in
	utils.go:163: Found agent metadata: {Cluster:ecs-functional-tests ContainerInstanceArn:0xc0422f5520 Version:Amazon ECS Agent - v1.23.0 (*UNKNOWN)}
	utils.go:184: Task definition: ecsftest-oom-container-windows-601ba04ca0a9e1d8536dadd3f48ea215:1
	utils.go:204: Started task: arn:aws:ecs:us-west-2:535959970326:task/cc2c3719-f779-4d6d-bbb9-c8c3a9ce2176
	functionaltests_windows_test.go:214: 
			Error Trace:	functionaltests_windows_test.go:214
			Error:      	Received unexpected error:
			            	Task terminal; will never reach RUNNING
			            	github.com/aws/amazon-ecs-agent/agent/functional_tests/util.(*TestTask).waitStatus.func1
			            		C:/Users/ADMINI~1/AppData/Local/Temp/amazon-ecs-agent/go/src/github.com/aws/amazon-ecs-agent/agent/functional_tests/util/utils.go:526
			            	runtime.goexit
			            		C:/Go/src/runtime/asm_amd64.s:2337
			Test:       	TestOOMContainer
			Messages:   	Expect task to be running
	utils.go:171: Preserving test dir for failed test C:\Users\ADMINI~1\AppData\Local\Temp\amazon-ecs-agent\ftest_temp\ecs_integ_testdata336277954
```

tracking here:  https://github.com/aws/amazon-ecs-agent/issues/1763

File: agent/engine/docker_image_manager.go [hunk 8]
```diff
@@ -335,7 +335,10 @@ func (imageManager *dockerImageManager) removeNonECSContainers(ctx context.Conte
 	}
 	var nonECSContainerRemoveAvailableIDs []string
 	for _, id := range nonECSContainersIDs {
-		response, _ := imageManager.client.InspectContainer(ctx, id, dockerclient.InspectContainerTimeout)
+		response, icErr := imageManager.client.InspectContainer(ctx, id, dockerclient.InspectContainerTimeout)
+		if icErr != nil {
+			seelog.Errorf("Error inspecting non-ECS container id: %s - %v", id, icErr)
```
Comment (haikuoliu) [inline agent/engine/docker_image_manager.go]: do we need to add a `continue` in this case since `response` will contain nothing? same below

Comment (adnxn) [reply agent/engine/docker_image_manager.go]: you're correct on both accounts, will update.

**Pull request merged**